packages/core/src/combinator/combine.ts

```diff
@@ -0,0 +1,108 @@
+/** @license MIT License (c) copyright 2010 original author or authors */
+
+import { map } from './transform'
+import { empty, containsCanonicalEmpty } from '../source/empty'
+import Pipe from '../sink/Pipe'
+import IndexSink, { IndexedValue } from '../sink/IndexSink' // eslint-disable-line no-unused-vars
+import { disposeAll, tryDispose } from '@most/disposable'
+import invoke from '../invoke'
+import { Stream, Sink, Scheduler, Disposable, Time } from '@most/types' // eslint-disable-line no-unused-vars
+import { ToStreamsArray } from './variadic' // eslint-disable-line no-unused-vars
+
+/**
+ * Combine latest events from two streams
+ * @param f function to combine most recent events
+ * @returns stream containing the result of applying f to the most recent
+ *  event of each input stream, whenever a new event arrives on any stream.
+ */
+export const combine = <A, B, C>(f: (a: A, b: B) => C, stream1: Stream<A>, stream2: Stream<B>): Stream<C> =>
+  combineArray(f, [stream1, stream2])
+
+/**
+* Combine latest events from all input streams
+* @param f function to combine most recent events
+* @param streams most recent events
+* @returns stream containing the result of applying f to the most recent
+*  event of each input stream, whenever a new event arrives on any stream.
+*/
+export const combineArray = <Args extends any[], R>(f: (...args: Args) => R, streams: ToStreamsArray<Args>): Stream<R> =>
+  streams.length === 0 || containsCanonicalEmpty(streams) ? empty()
+    : streams.length === 1 ? map(f as any, streams[0])
+      : new Combine(f as any, streams)
+
+class Combine<A, B> {
```

Comment:
This was my first move, but then I realized it was just too tedious with no profit because of the structural nature of type system. Even more TS still needs the types on `run` method, it just cannot pick them from the interface.

Reply:
Anyway I can add annotations for the sake of codestyle

Reply:
Thanks.  I'm actually a fan of structural types and "lazy" type compatibility in many cases.  In this case, though, since `Stream` is such an integral type, I think there are a few subtle, but important advantages of being explicit:

1. For future readers & refactorers, it makes it obvious if a class is a Stream, without having to scan the code to see if it implements a properly typed run method.
2. It sets an example for developers who create new (or update existing) 3rd-party mostjs libs, and encourages them to `implements Stream<A>` also, which will tend to find type errors earlier than relying on structural types.  Example: if `@most/types` were to release a new major version with a breaking change to the Stream type, IDEs would flag it immediately at the class declaration site, rather than some other place (like in a unit test) in the code.

Reply:
done